README.md

[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/wrapr)](https://cran.r-project.org/package=wrapr)
[![status](https://tinyverse.netlify.com/badge/wrapr)](https://CRAN.R-project.org/package=wrapr)

<!-- README.md is generated from README.Rmd. Please edit that file -->

[`wrapr`](https://winvector.github.io/wrapr/) is an
[`R`](https://cran.r-project.org) package that supplies powerful tools
for writing and debugging `R` code.

![](https://github.com/WinVector/wrapr/raw/master/tools/wraprs.png)

## Introduction

Primary `wrapr` services include:

- [`%.>%`](https://winvector.github.io/wrapr/reference/dot_arrow.html)
  (dot arrow pipe)
- [`unpack`](https://winvector.github.io/wrapr/reference/unpack.html)/[`to`](https://winvector.github.io/wrapr/reference/to.html)
  (assign to multiple values)
- [`as_named_list`](https://winvector.github.io/wrapr/reference/as_named_list.html)
  (build up a named list quickly)
- [`build_frame()`](https://winvector.github.io/wrapr/reference/build_frame.html)
  /
  [`draw_frame()`](https://winvector.github.io/wrapr/reference/draw_frame.html)
  ( `data.frame` builders and formatters )
- [`bc()`](https://win-vector.com/2021/02/05/introducing-wraprbc/)
  (blank concatenate)
- [`qc()`](https://winvector.github.io/wrapr/reference/qc.html) (quoting
  concatenate)
- [`:=`](https://winvector.github.io/wrapr/reference/named_map_builder.html)
  (named map builder)
- [`%?%`](https://winvector.github.io/wrapr/reference/coalesce.html)
  (coalesce)
- [`%.|%`](https://winvector.github.io/wrapr/reference/reduceexpand.html)
  (reduce/expand args)
- [`uniques()`](https://winvector.github.io/wrapr/reference/uniques.html)
  (safe `unique()` replacement)
- [`partition_tables()`](https://winvector.github.io/wrapr/reference/partition_tables.html)
  /
  [`execute_parallel()`](https://winvector.github.io/wrapr/reference/execute_parallel.html)
- [`DebugFnW()`](https://winvector.github.io/wrapr/reference/DebugFnW.html)
  (function debug wrappers)
- [`λ()`](https://winvector.github.io/wrapr/reference/defineLambda.html)
  (anonymous function builder)
- [`let()`](https://winvector.github.io/wrapr/reference/let.html) (let
  block)
- [`evalb()`/`si()`](https://winvector.github.io/wrapr/articles/bquote.html)
  (evaluate with `bquote` / string interpolation)
- [`sortv()`](https://winvector.github.io/wrapr/reference/sortv.html)
  (sort a data.frame by a set of columns).
- [`stop_if_dot_args()`](https://win-vector.com/2018/02/22/r-tip-force-named-arguments/)
  (check for unexpected arguments)

``` r
library(wrapr)
packageVersion("wrapr")
 #  [1] '2.1.0'
date()
 #  [1] "Sat Aug 19 09:06:13 2023"
```

## [`%.>%` (dot pipe or dot arrow)](https://winvector.github.io/wrapr/articles/dot_pipe.html)

`%.>%` dot arrow pipe is a pipe with intended semantics:

> “`a %.>% b`” is to be treated *approximately* as if the user had
> written “`{ . <- a; b };`” with “`%.>%`” being treated as
> left-associative.

Other `R` pipes include
[`magrittr`](https://CRAN.R-project.org/package=magrittr) and
[`pipeR`](https://CRAN.R-project.org/package=pipeR).

The following two expressions should be equivalent:

``` r
cos(exp(sin(4)))
 #  [1] 0.8919465

4 %.>% sin(.) %.>% exp(.) %.>% cos(.)
 #  [1] 0.8919465
```

The notation is quite powerful as it treats pipe stages as expression
parameterized over the variable “`.`”. This means you do not need to
introduce functions to express stages. The following is a valid
dot-pipe:

``` r
1:4 %.>% .^2 
 #  [1]  1  4  9 16
```

The notation is also very regular as we show below.

``` r
1:4 %.>% sin
 #  [1]  0.8414710  0.9092974  0.1411200 -0.7568025
1:4 %.>% sin(.)
 #  [1]  0.8414710  0.9092974  0.1411200 -0.7568025
1:4 %.>% base::sin
 #  [1]  0.8414710  0.9092974  0.1411200 -0.7568025
1:4 %.>% base::sin(.)
 #  [1]  0.8414710  0.9092974  0.1411200 -0.7568025

1:4 %.>% function(x) { x + 1 }
 #  [1] 2 3 4 5
1:4 %.>% (function(x) { x + 1 })
 #  [1] 2 3 4 5

1:4 %.>% { .^2 } 
 #  [1]  1  4  9 16
1:4 %.>% ( .^2 )
 #  [1]  1  4  9 16
```

Regularity can be a *big* advantage in teaching and comprehension.
Please see [“In Praise of Syntactic
Sugar”](https://win-vector.com/2017/07/07/in-praise-of-syntactic-sugar/)
for more details. Some formal documentation can be found
[here](https://github.com/WinVector/wrapr/blob/master/extras/wrapr_pipe.pdf).

<ul>
<li>
Some obvious “dot-free”” right-hand sides are rejected. Pipelines are
meant to move values through a sequence of transforms, and not just for
side-effects. Example: `5 %.>% 6` deliberately stops as `6` is a
right-hand side that obviously does not use its incoming value. This
check is only applied to values, not functions on the right-hand side.
</li>
<li>
Trying to pipe into a an “zero argument function evaluation expression”
such as `sin()` is prohibited as it looks too much like the user
declaring `sin()` takes no arguments. One must pipe into either a
function, function name, or an non-trivial expression (such as
`sin(.)`). A useful error message is returned to the user:
`wrapr::pipe does not allow direct piping into a no-argument function call expression (such as "sin()" please use sin(.))`.
</li>
<li>
Some reserved words can not be piped into. One example is
`5 %.>% return(.)` is prohibited as the obvious pipe implementation
would not actually escape from user functions as users may intend.
</li>
<li>
Obvious de-references (such as `$`, `::`, `@`, and a few more) on the
right-hand side are treated performed (example: `5 %.>% base::sin(.)`).
</li>
<li>
Outer parenthesis on the right-hand side are removed (example:
`5 %.>% (sin(.))`).
</li>
<li>
Anonymous function constructions are evaluated so the function can be
applied (example: `5 %.>% function(x) {x+1}` returns 6, just as
`5 %.>% (function(x) {x+1})(.)` does).
</li>
<li>
Checks and transforms are not performed on items inside braces (example:
`5 %.>% { function(x) {x+1} }` returns `function(x) {x+1}`, not 6).
</li>
<li>
The dot arrow pipe has S3/S4 dispatch (please see ). However as the
right-hand side of the pipe is normally held unevaluated, we don’t know
the type except in special cases (such as the rigth-hand side being
referred to by a name or variable). To force the evaluation of a pipe
term, simply wrap it in `.()`.
</li>
</ul>

The dot pipe is also user configurable through standard `S3`/`S4`
methods.

The dot pipe has been [formally written up in the R
Journal](https://journal.r-project.org/archive/2018/RJ-2018-042/index.html).

    @article{RJ-2018-042,
      author = {John Mount and Nina Zumel},
      title = {{Dot-Pipe: an S3 Extensible Pipe for R}},
      year = {2018},
      journal = {{The R Journal}},
      url = {https://journal.r-project.org/archive/2018/RJ-2018-042/index.html}
    }

## [`unpack`](https://winvector.github.io/wrapr/reference/unpack.html)/[`to`](https://winvector.github.io/wrapr/reference/to.html) multiple assignments

Unpack a named list into the current environment by name (for a
positional based multiple assignment operator please see
[`zeallot`](https://CRAN.R-project.org/package=zeallot), for another
named base multiple assigment please see
[`vadr::bind`](https://github.com/crowding/vadr/blob/master/R/bind.R)).

``` r
d <- data.frame(
  x = 1:9,
  group = c('train', 'calibrate', 'test'),
  stringsAsFactors = FALSE)

unpack[
  train_data = train,
  calibrate_data = calibrate,
  test_data = test
  ] := split(d, d$group)

knitr::kable(train_data)
```

|     |   x | group |
|:----|----:|:------|
| 1   |   1 | train |
| 4   |   4 | train |
| 7   |   7 | train |

## [`as_named_list`](https://winvector.github.io/wrapr/reference/as_named_list.html)

Build up named lists. [Very convenient for managing workspaces when used
with used with
`unpack`/`to`.](https://win-vector.com/2020/01/21/using-unpack-to-manage-your-r-environment/)

``` r
as_named_list(train_data, calibrate_data, test_data)
 #  $train_data
 #    x group
 #  1 1 train
 #  4 4 train
 #  7 7 train
 #  
 #  $calibrate_data
 #    x     group
 #  2 2 calibrate
 #  5 5 calibrate
 #  8 8 calibrate
 #  
 #  $test_data
 #    x group
 #  3 3  test
 #  6 6  test
 #  9 9  test
```

## [`build_frame()`](https://winvector.github.io/wrapr/reference/build_frame.html) / [`draw_frame()`](https://winvector.github.io/wrapr/reference/draw_frame.html)

[`build_frame()`](https://winvector.github.io/wrapr/reference/build_frame.html)
is a convenient way to type in a small example `data.frame` in natural
row order. This can be very legible and saves having to perform a
transpose in one’s head.
[`draw_frame()`](https://winvector.github.io/wrapr/reference/draw_frame.html)
is the complimentary function that formats a given `data.frame` (and is
a great way to produce neatened examples).

``` r
x <- build_frame(
   "measure"                   , "training", "validation" |
   "minus binary cross entropy", 5         , -7           |
   "accuracy"                  , 0.8       , 0.6          )
print(x)
 #                       measure training validation
 #  1 minus binary cross entropy      5.0       -7.0
 #  2                   accuracy      0.8        0.6
str(x)
 #  'data.frame':   2 obs. of  3 variables:
 #   $ measure   : chr  "minus binary cross entropy" "accuracy"
 #   $ training  : num  5 0.8
 #   $ validation: num  -7 0.6
cat(draw_frame(x))
 #  x <- wrapr::build_frame(
 #     "measure"                     , "training", "validation" |
 #       "minus binary cross entropy", 5         , -7           |
 #       "accuracy"                  , 0.8       , 0.6          )
```

## [`qc()` (quoting concatenate)](https://winvector.github.io/wrapr/reference/qc.html)

`qc()` is a quoting variation on `R`’s concatenate operator `c()`. This
code such as the following:

``` r
qc(a = x, b = y)
 #    a   b 
 #  "x" "y"

qc(one, two, three)
 #  [1] "one"   "two"   "three"
```


NEWS.md

# wrapr 2.1.0 2023-08-19

 * Work around https://github.com/r-lib/roxygen2/issues/1491
 * For captured_dots note in CRAN checks
 
# wrapr 2.0.9 2022-01-26

 * Document qc() limitations (ref: https://github.com/WinVector/wrapr/issues/15#issue-1046032267 ).
 * Add sx() for https://github.com/WinVector/wrapr/issues/15#issue-1046032267 use case.

# wrapr 2.0.8 2021-06-10

 * Make calls to order stricter (no names, and convert data.frames), 
   see https://stat.ethz.ch/pipermail/r-devel/2020-May/079500.html 
   and https://stat.ethz.ch/pipermail/r-package-devel/2021q1/006550.html
 * Add international character tests for bc().
   
# wrapr 2.0.7 2021-02-03

 * Add bc() blank-separated parsing, suggested by Emil Erik Pula Bellamy Begtrup-Bright \url{https://github.com/WinVector/wrapr/issues/12}.
 * Fix strsplit_capture empty-return type.

# wrapr 2.0.6 2020-12-06

 * Bump version.
 * Add pack().

# wrapr 2.0.5 2020-11-13

 * More ... checks.
 * Better error message for extra comman in unpack.
 * Remove RUnit suggest / test function.

# wrapr 2.0.4 2020-10-16

 * Move to tinytest.

# wrapr 2.0.2 2020-08-10

 * dot-function (f.).
 * character-mode for mk_formula.
 * work around refs such as `\link[utils]{dump.frames}` no longer working with Ben Boker's suggestion `\link[utils:debugger]{dump.frames}` on R-pkg-devel
 * URLs to https where possible.

# wrapr 2.0.0 2020-03-27

 * Allow as_named_list to choose new names.
 * Update references.
 * Stricter tests on unpack/to.
 * Stricter wrapr-pipe value checks.
 * := for names.
 * := for to/unpack.
 * Don't use := for anonymous function construction.
 * Update evalb, and bquote uses including adding .(-) notation.
 * Remove some obsolete methods such as CapturePipeine, as_dot_fn, UnaryFunctions/ApplyTo, and locum.
 * Make pipe_impl public, and document more.
 * Vectorize string interpolation and add operator versions.
 * Fix description.

# wrapr 1.9.6 2020-01-26

 * Add dotpipe_eager_eval_*.
 * Add on unpack method.

# wrapr 1.9.5 2020-01-18

 * remove q= from sQuote to, as that feature wasn't added until newer Rs.

# wrapr 1.9.4 2020-01-16

 * tolerance on check_equiv_frames
 * remove graphics dependency
 * improve ... check message.

# wrapr 1.9.3 2019-11-03

 * := for factors and logical

# wrapr 1.9.2 2019-10-13

 * Fix dev version at request of CRAN, appears to be instability in capturing ... and passing to bquote(). https://github.com/WinVector/wrapr/blob/master/extras/wrapr_dev_CRAN_issue_2019_10_13.md

# wrapr 1.9.1 2019-10-05

 * Fix locum composite line printing.
 * Document .()-escaping in pipe.

# wrapr 1.9.0 2019-09-28

 * Add locum ( https://github.com/WinVector/wrapr/blob/master/Examples/Locum/Locum.md ).

# wrapr 1.8.9 2019-07-24

 * More unit tests.
 * Adjust license.

# wrapr 1.8.8 2019-07-06

 * More work with mk_formula().

# wrapr 1.8.7 2019-05-23

 * Better sinterp semantics (vectorize over values, not template).
 * Looser checks on qchar_frame().

# wrapr 1.8.6 2019-04-02

  * Turn off radix sort to work around https://github.com/WinVector/wrapr/issues/9
  * Make draw_frame() resistent to operator driven re-indentation.
  * Add draw_framec().

# wrapr 1.8.5 2019-03-24

  * More tests.
  * Be more careful with formals().
  * Add frame checking utils.
  * Doc fixes.

# wrapr 1.8.4 2019-02-19

  * Add run_packages_tests() and run_wrapr_tests().
  * Move to RUnit tests.
  * More args to clean_fit.

# wrapr 1.8.3 2019-01-29

 * Add string strsplit_capture() and dotsubs().
 * better bquote_function() and evalb() examples.
 * Allow comparisions in mk_formula(), and general improvements.

# wrapr 1.8.2 2019-01-04

 * Add paste(class(), collapse=" ").
 * Add seqi().
 * Add evalb().

# wrapr 1.8.1 2018-12-13

 * Fix as_fn() environment assignment.

# wrapr 1.8.0 2018-12-11

 * Add generic unary functions.
 * Add .() pipe escaping.
 * Add clean lm() and glm() fitters.
 * Add split_at_brace_pairs.
 * Documentation fixes.
 * Error msg fix.
 * Add %p%, and %.%.
 
# wrapr 1.7.0 2018-11-15

 * bquote enable qc() and other quoting methods.
 * Preserve NA types in draw_frame.
 * Add VectorizeM, vapplym, lapplym.
 * Add bquote_function().
 * Export underbar version of apply_left_default to get non-S3 version of code.
 * Better string concat example.
 * Add %c% and %qc%.
 * Allow qc() to call c().

# wrapr 1.6.3 2018-10-03

 * Make sure parent.frame() is unambiguosly resovled (force()).
 * qe(), qae(), qs() now return character vectors instead of lists.
 * Add psagg().
 * Add grepv().
 * More examples in SubstitutionModes vignette (show we don't need special pairlist case).
 * Allow no-intercept version of formula.
 
# wrapr 1.6.2 2018-09-10

 * Add mk_formula().
 * Documentation fixes.
 * Add %in_block% operator notation.
 * Add orderv().
 
# wrapr 1.6.1 2018-08-21

 * More restrictive apply_right_S4.
 * Clear methods note.
 
# wrapr 1.6.0 2018-08-01
 
 * S4 dispatch apply_right_S4.

R/cf.R

is_infix <- function(vi) {
  vi <- as.character(vi)
  if(nchar(vi)<=0) {
    return(FALSE)
  }
  if(substr(vi,1,1)=="`") {
    vi <- substr(vi,2,nchar(vi)-1)
  }
  if(nchar(vi)<=0) {
    return(FALSE)
  }
  if(substr(vi,1,1)=="%") {
    return(TRUE)
  }
  syms <- c("::", "$", "@", "^", ":",
            "*", "/", "+", "-",
            ">", ">=", "<", "<=",  "==", "!=",
            "&", "&&",
            "|", "||",
            "~",
            "->",  "->>",
            "=",
            "<-", "<<-")
  if(vi %in% syms) {
    return(TRUE)
  }
  return(FALSE)
}

# convert a list to a vector without losing type/class info
# lst is a list of scalars
to_vector <- function(lst) {
  n <- length(lst)
  if(n<1) {
    return(logical(0))
  }
  vec <- rep(lst[[1]], n)
  for(i in seqi(2,n)) {
    vec[[i]] <- lst[[i]]
  }
  vec
}

#' Build a data.frame from the user's description.
#'
#' A convenient way to build a data.frame in legible transposed form.  Position of
#' first "|" (or other infix operator) determines number of columns
#' (all other infix operators are aliases for ",").
#' Names are de-referenced.
#'
#' @param ... cell names, first infix operator denotes end of header row of column names.
#' @param cf_eval_environment environment to evaluate names in.
#' @return character data.frame
#'
#' @seealso \code{\link{draw_frame}}, \code{\link{qchar_frame}}
#'
#' @examples
#'
#' tc_name <- "training"
#' x <- build_frame(
#'    "measure",                   tc_name, "validation" |
#'    "minus binary cross entropy",      5, -7           |
#'    "accuracy",                      0.8, 0.6          )
#' print(x)
#' str(x)
#' cat(draw_frame(x))
#'
#' build_frame(
#'   "x" |
#'   -1  |
#'   2   )
#'
#' @export
#'
build_frame <- function(..., cf_eval_environment = parent.frame()) {
  v <- as.list(substitute(list(...)))
  lengths <- vapply(
    v,
    function(vi) {nchar(paste(as.character(vi), collapse = ''))},
    numeric(1))
  if(any(lengths <= 0)) {
    stop("empty entry, this is often caused by an extra comma")
  }
  v <- lapply(seqi(2, length(v)), function(i) {v[[i]]})
  force(cf_eval_environment)
  lv <- length(v)
  # inspect input
  if(lv<1) {
    return(data.frame())
  }
  # unpack
  unpack_val <- function(vi) {
    if(length(vi)<=0) {
      stop("wrapr::build_frame unexpected NULL/empty element")
    }
    if(is.name(vi)) {
      viv <- cf_eval_environment[[as.character(vi)]]
      if(is.name(viv)) {
        stop(paste("wrapr::build_frame name",
                   vi,
                   "resolved to another name:",
                   viv))
      }
      if(is.call(viv)) {
        stop(paste("wrapr::build_frame name",
                   vi,
                   "resolved to call",
                   viv))
      }
      if(length(viv)<=0) {
        stop(paste("wrapr::build_frame name",
                   vi,
                   "resolved to NULL"))
      }
      vi <- viv
    }
    if(is.call(vi)) {
      if((length(vi)==3) && (is_infix(vi[[1]]))) {
        vi <- list(unpack_val(vi[[2]]),
                   as.name("sep"),
                   unpack_val(vi[[3]]))
      } else {
        viv <- eval(vi,
                   envir = cf_eval_environment,
                   enclos = cf_eval_environment)
        if(is.name(viv)) {
          stop(paste("wrapr::build_frame eval",
                     vi,
                     "resolved to another name:",
                     viv))
        }
        if(length(viv)<=0) {
          stop(paste("wrapr::build_frame eval",
                     vi,
                     "resolved to NULL"))
        }
        vi <- viv
      }
    }
    Reduce(c, lapply(vi, as.list))
  }
  vu <- lapply(v, unpack_val)
  vu <- Reduce(c, lapply(vu, as.list))
  ncol <- length(vu)
  if(ncol<1) {
    stop("wrapr::build_frame() zero columns")
  }
  is_name <- vapply(vu, is.name, logical(1))
  if(any(is_name)) {
    ncol <- which(is_name)[[1]]-1
    vu <- vu[!is_name] # filter out names
  }
  nrow <- (length(vu)/ncol) - 1
  if(abs(nrow - round(nrow))>0.1) {
    stop("wrapr::build_frame confused as to cell count")
  }
  if(nrow<=0) {
    fr <- data.frame(x = logical(0))
    colnames(fr) <- as.character(vu[[1]])
    if(ncol>1) {
      for(i in 2:ncol) {
        ci <- as.character(vu[[i]])
        fr[[ci]] <- logical(0)
      }
    }
  } else {
    seq <- seq_len(nrow)*ncol
    fr <- data.frame(x = to_vector(vu[seq + 1]),
                     stringsAsFactors = FALSE)
    colnames(fr) <- as.character(vu[[1]])
    if(ncol>1) {
      for(i in 2:ncol) {
        ci <- as.character(vu[[i]])
        fr[[ci]] <-  to_vector(vu[seq + i])
      }
    }
  }
  rownames(fr) <- NULL
  fr
}


#' Render a simple data.frame in build_frame format.
#'
#' @param x data.frame (with atomic types).
#' @param ... not used for values, forces later arguments to bind by name.
#' @param time_format character, format for "POSIXt" classes.
#' @param formatC_options named list, options for formatC()- used on numerics.
#' @param adjust_for_auto_indent integer additional after first row padding
#' @return character
#'
#' @seealso \code{\link{build_frame}},  \code{\link{qchar_frame}}
#'
#' @examples
#'
#' tc_name <- "training"
#' x <- build_frame(
#'   "measure"                   , tc_name, "validation", "idx" |
#'   "minus binary cross entropy", 5      , 7           , 1L    |
#'   "accuracy"                  , 0.8    , 0.6         , 2L    )
#' print(x)
#' cat(draw_frame(x))
#'
#'
#' @export
#'
draw_frame <- function(x,
                       ...,
                       time_format = "%Y-%m-%d %H:%M:%S",
                       formatC_options = list(),
                       adjust_for_auto_indent = 2) {
  wrapr::stop_if_dot_args(substitute(list(...)), "wrapr::draw_frame")
  formatC_args = list(digits = NULL,
                      width = NULL,
                      format = NULL,
                      flag = "",
                      mode = NULL,
                      big.mark = "",
                      big.interval = 3L,
                      small.mark = "",
                      small.interval = 5L,
                      decimal.mark = getOption("OutDec"),
                      preserve.width = "individual",
                      zero.print = NULL,
                      drop0trailing = FALSE)
  x_s <- substitute(x)
  for(oi in names(formatC_options)) {
    formatC_args[[oi]] <- formatC_options[[oi]]
  }
  if(!is.data.frame(x)) {
    stop("draw_frame x needs to be a data.frame")
  }
  res <- "wrapr::build_frame()"
  nrow <- nrow(x)
  ncol <- ncol(x)
  if((nrow>=1) && (ncol<1)) {
    stop("wrapr::draw_frame bad input: no columns, but has rows")
  }
  qts <- function(v) {
    # wayts to quote: dput(), shQuote(), deparse()
    vapply(as.character(v),
           function(vi) {
             deparse(vi)
           },
           character(1))
  }
  if((nrow<1) || (ncol<1)) {
    if(ncol>=1) {
      res <- paste(qts(colnames(x)), collapse = ", ")
      res <- paste0("wrapr::build_frame(", res, ")")
    }
  } else {
    # convert to character matrix
    xq <- x

    for(ci in colnames(x)) {
      if("Date" %in% class(x[[ci]])) {
        xq[[ci]] <- paste0("as.Date(\"",
                           format(x[[ci]]),
                           "\")")
        xq[[ci]][is.na(x[[ci]])] <- "NA_real_"
      } else if("POSIXt" %in% class(x[[ci]])) {
        # round tripping a time correctly through R is so unlikely, just fall back to string
        xq[[ci]] <- paste0("\"",
                           format(x[[ci]], time_format),
                           "\"")
        xq[[ci]][is.na(x[[ci]])] <- "NA_character_"
      } else if(is.character(x[[ci]]) || is.factor(x[[ci]])) {
        xq[[ci]] <- qts(as.character(x[[ci]]))
        xq[[ci]][is.na(x[[ci]])] <- "NA_character_"
      } else if(is.integer(x[[ci]])) {
        xq[[ci]] <- paste0(format(x[[ci]], scientific = FALSE), "L")
        xq[[ci]][is.na(x[[ci]])] <- "NA_integer_"
      } else if(is.numeric(x[[ci]])) {
        xq[[ci]] <- formatC(x[[ci]],
                            digits = formatC_args$digits,
                            width =  formatC_args$width,
                            format =  formatC_args$format,
                            flag =  formatC_args$flag,
                            mode =  formatC_args$mode,
                            big.mark =  formatC_args$big.mark,
                            big.interval =  formatC_args$big.interval,
                            small.mark =  formatC_args$small.mark,
                            small.interval =  formatC_args$small.interval,
                            decimal.mark =  formatC_args$decimal.mark,
                            preserve.width =  formatC_args$preserve.width,
                            zero.print =  formatC_args$zero.print,
                            drop0trailing =  formatC_args$drop0trailing)
        xq[[ci]][is.na(x[[ci]])] <- "NA_real_"
      } else if(is.logical(x[[ci]])) {
        xq[[ci]] <- as.character(x[[ci]])
        xq[[ci]][is.na(x[[ci]])] <- "NA"
      } else {
        xq[[ci]] <- as.character(x[[ci]])
        xq[[ci]][is.na(x[[ci]])] <- "NA"
      }
    }
    xm <- as.matrix(xq)
    xm <- matrix(data = as.character(xm),
                 nrow = nrow, ncol = ncol)
    # convert header to values
    xm <- rbind(matrix(data = qts(colnames(x)),
                       nrow = 1, ncol = ncol),
                xm)
    # compute padding
    widths <- nchar(xm)
    widths[is.na(as.numeric(widths))] <- 2
    colmaxes <- matrix(data = apply(widths, 2, max),
                       nrow = nrow+1, ncol = ncol,
                       byrow = TRUE)
    padlens <- colmaxes - widths
    pads <- matrix(data = vapply(padlens,
                                 function(vi) {
                                   paste(rep(' ', vi), collapse = '')
                                 }, character(1)),
                   nrow = nrow+1, ncol = ncol)
    # get intermediates
    seps <- matrix(data = ", ",
                   nrow = nrow+1, ncol = ncol)
    seps[, ncol] <- " |"
    seps[nrow+1, ncol] <- " )"
    # format
    fmt <- matrix(data = paste0(xm, pads, seps),
                  nrow = nrow+1, ncol = ncol)
    if(adjust_for_auto_indent>0) {
      pad <- paste(rep(" ", adjust_for_auto_indent), collapse = "")
      fmt[1, 1] <- gsub(", $", paste0(pad, ", "), fmt[1, 1])
      for(i in seqi(2, nrow(fmt))) {
        fmt[i, 1] <- paste0(pad, fmt[i, 1])
      }
    }
    rlist <- vapply(seq_len(nrow+1),
                    function(i) {
                      paste(fmt[i, , drop=TRUE], collapse = '')
                    }, character(1))
    rlist <- paste0("   ", rlist)
    res <- paste(rlist, collapse = "\n")
    res <- paste0("wrapr::build_frame(\n", res, "\n")
  }
  if(is.name(x_s)) {
    res <- paste0(as.character(x_s), " <- ", res)
  }
  res
}



#' Build a quoted data.frame.
#'
#' A convenient way to build a character data.frame in legible transposed form.  Position of
#' first "|" (or other infix operator) determines number of columns
#' (all other infix operators are aliases for ",").
#' Names are treated as character types.
#'
#' qchar_frame() uses bquote() .() quasiquotation escaping notation.  Because of this using dot
#' as a name in some places may fail if the dot looks like a function call.
#'
#' @param ... cell names, first infix operator denotes end of header row of column names.
#' @return character data.frame
#'
#' @seealso \code{\link{draw_frame}}, \code{\link{build_frame}}
#'
#' @examples
#'
#' loss_name <- "loss"
#' x <- qchar_frame(
#'    measure,                      training,     validation |
#'    "minus binary cross entropy", .(loss_name), val_loss   |
#'    accuracy,                     acc,          val_acc    )
#' print(x)
#' str(x)
#' cat(draw_frame(x))
#'
#' qchar_frame(
#'   x |
#'   1 |
#'   2 ) %.>% str(.)
#'
#' @export
#'
qchar_frame <- function(...) {
  env <- parent.frame()
  v <- do.call(bquote, list(substitute(alist(...)),
                            where = env),
               envir = env)
  v <- lapply(seqi(2, length(v)), function(i) {v[[i]]})
  lv <- length(v)
  if(lv<1) {
    return(data.frame())
  }
  lengths <- vapply(
    v,
    function(vi) {nchar(paste(as.character(vi), collapse = ''))},
    numeric(1))
  if(any(lengths <= 0)) {
    stop("empty entry, this is often caused by an extra comma")
  }
  # inspect input
  cls <- vapply(v, class, character(1))
  if(length(setdiff(cls, c("call", "character", "name", "numeric", "integer", "logical", "factor")))>0) {
    stop("wrapr::qchar_frame expects only strings, names, literals, operators, and commas")
  }
  if(sum(cls=="call") < 1) {
    # no rows case
    fr <- data.frame(x = character(0),
                     stringsAsFactors = FALSE)
    colnames(fr) <- as.character(v[[1]])
    if(lv>1) {
      for(i in 2:lv) {
        fr[[as.character(v[[i]])]] <- character(0)
      }
    }
    rownames(fr) <- NULL
    return(fr)
  }
  ncol <- match("call", cls)
  # unpack
  unpack_val <- function(vi) {
    if(length(vi)<=0) {
      stop("wrapr::qchar_frame unexpected NULL/empty element")
    }
    if(is.call(vi)) {
      if((length(vi)!=3) || (!is_infix(vi[[1]]))) {
        stop(paste("wrapr::qchar_frame unexpected operator", vi[[1]]))
      }
      vi <- lapply(as.list(vi)[-1], unpack_val)
    }
    as.character(unlist(vi))
  }
  vu <- lapply(v, unpack_val)
  vu <- unlist(vu)
  ncell <- length(vu)
  nrow <- ncell/ncol
  if(abs(nrow - round(nrow))>0.1) {
    stop("wrapr::qchar_frame confused as to cell count (this can be an extra comma or mis-placed separator)")
  }
  fr <- as.data.frame(matrix(data = vu[-seq_len(ncol)],
                             ncol=ncol,
                             byrow = TRUE),
                      stringsAsFactors = FALSE)
  colnames(fr) <- vu[seq_len(ncol)]
  rownames(fr) <- NULL
  fr
}


#' Render a simple data.frame in qchar_frame format.
#'
#' @param x data.frame (with character types).
#' @param ... not used for values, forces later arguments to bind by name.
#' @param unquote_cols character, columns to elide quotes from.
#' @param adjust_for_auto_indent integer additional after first row padding.
#' @return character
#'
#' @seealso \code{\link{build_frame}},  \code{\link{qchar_frame}}
#'
#' @examples
#'
#' controlTable <- wrapr::qchar_frame(
#'   "flower_part", "Length"     , "Width"     |
#'     "Petal"    , Petal.Length , Petal.Width |
#'     "Sepal"    , Sepal.Length , Sepal.Width )
#' cat(draw_framec(controlTable, unquote_cols = qc(Length, Width)))
#'
#'
#' @export
#'
draw_framec <- function(x,
                        ...,
                        unquote_cols = character(0),
                        adjust_for_auto_indent = 2) {
  wrapr::stop_if_dot_args(substitute(list(...)), "wrapr::draw_framec")
  x_s <- substitute(x)
  if(!is.data.frame(x)) {
    stop("wrapr::draw_framec x needs to be a data.frame")
  }
  res <- "wrapr::qchar_frame()"
  nrow <- nrow(x)
  ncol <- ncol(x)
  if((nrow>=1) && (ncol<1)) {
    stop("wrapr::draw_framec bad input: no columns, but has rows")
  }
  qts <- function(v) {
    # wayts to quote: dput(), shQuote(), deparse()
    vapply(as.character(v),
           function(vi) {
             deparse(vi)
           },
           character(1))
  }
  if((nrow<1) || (ncol<1)) {
    if(ncol>=1) {
      res <- paste(qts(colnames(x)), collapse = ", ")
      res <- paste0("wrapr::qchar_frame(", res, ")")
    }
  } else {
    # convert to character matrix
    xq <- x
    for(ci in colnames(x)) {
      if(ci %in% unquote_cols) {
        xq[[ci]] <- as.character(x[[ci]])
      } else {
        xq[[ci]] <- qts(as.character(x[[ci]]))
      }
      xq[[ci]][is.na(x[[ci]])] <- "NA_character_"
    }
    xm <- as.matrix(xq)
    xm <- matrix(data = as.character(xm),
                 nrow = nrow, ncol = ncol)
    # convert header to values
    xm <- rbind(matrix(data = qts(colnames(x)),
                       nrow = 1, ncol = ncol),
                xm)
    # compute padding
    widths <- nchar(xm)
    widths[is.na(as.numeric(widths))] <- 2
    colmaxes <- matrix(data = apply(widths, 2, max),
                       nrow = nrow+1, ncol = ncol,
                       byrow = TRUE)
    padlens <- colmaxes - widths
    pads <- matrix(data = vapply(padlens,
                                 function(vi) {
                                   paste(rep(' ', vi), collapse = '')
                                 }, character(1)),
                   nrow = nrow+1, ncol = ncol)
    # get intermediates
    seps <- matrix(data = ", ",
                   nrow = nrow+1, ncol = ncol)
    seps[, ncol] <- " |"
    seps[nrow+1, ncol] <- " )"
    # format
    fmt <- matrix(data = paste0(xm, pads, seps),
                  nrow = nrow+1, ncol = ncol)
    if(adjust_for_auto_indent>0) {
      pad <- paste(rep(" ", adjust_for_auto_indent), collapse = "")
      fmt[1, 1] <- gsub(", $", paste0(pad, ", "), fmt[1, 1])
      for(i in wrapr::seqi(2, nrow(fmt))) {
        fmt[i, 1] <- paste0(pad, fmt[i, 1])
      }
    }
    rlist <- vapply(seq_len(nrow+1),
                    function(i) {
                      paste(fmt[i, , drop=TRUE], collapse = '')
                    }, character(1))
    rlist <- paste0("   ", rlist)
    res <- paste(rlist, collapse = "\n")
    res <- paste0("wrapr::qchar_frame(\n", res, "\n")
  }
  if(is.name(x_s)) {
    res <- paste0(as.character(x_s), " <- ", res)
  }
  res
}

